### ISSUE TITLE
ngtests: test loading of cached metadata

### ISSUE THREAD

**Issue by sechkova** [theupdateframework/tuf#1681]
**Description of issue or feature request**:
Taken from @jku's [comment](https://github.com/theupdateframework/python-tuf/pull/1636#issuecomment-953669058) that summarizes it very well,  tests are needed which:

- check that expected metadata is read from cache the expected number of times
     - wrapping open() should work fine here (see tests in #1593)
-  check if cached metadata is used to verify new metadata downloaded from remote
      - trickier but I suppose we can rely on negative tests here: tests that are expected to fail on that verification

**Comment by ivanayov** [comment]
Can I take this, please?

**Comment by jku** [comment]
I think teodoras current approach of mocking open (in #1593) is pretty good: just implementing this in the relevant tests  is probably fine (so no test utilities needed).

The question is, which tests should test metadata file opens:
* the test in #1593 should be extended so it checks files opened during refresh() as well
* any other tests needed or is this one enough?
 
Maybe this single test is enough? @sechkova opinions?

**Comment by sechkova** [comment]
What I imagined while writing the issue is potentially having a class that tests all weird cases of loading cached metadata. This would mean at least:

Step 1:  make a successful update of valid metadata which stores it in cache (by calling preferably Updater.get_targetinfo or in some special case Updater.refresh)
Step 2: Create a new updater and perform a second update while the metadata is already stored in cache (metadata dir)
Step 3:
- test that a valid timestamp is loaded from cache (and is used to verify the new timestamp)
- test that if a valid snapshot is loaded from cache a new snapshot is not downloaded
- test that  if a valid targets is loaded from cache a new targets is not downloaded

Step 4:
- test that an expired timestamp when loaded from cache is not stored as final but is used to verify the new timestamp
- test that an expired snapshot when loaded from cache  is not stored as final but is used to verify the new downloaded snapshot
- ~test that an expired targets when loaded from cache  is not stored as final but is used to verify the new downloaded targets~
- expired targets is invalid and should be neither loaded, stored nor used.

Step 5:
- test that an invalid (incorrectly signed) timestamp when loaded from cache IS NOT used to verify the new timestamp
- test that an invalid (incorrectly signed) snapshot when loaded from cache IS NOT used to verify the new timestamp
- test that an invalid (incorrectly signed) targets when loaded from cache IS NOT used ~to verify the new targets~

Step 3 can be based on the test in #1593 but I think a separate test case will be more clear. I haven't thought out well steps 4 and 5 but I'm open for suggestions. Maybe start with 3 and re-scope the issue accordingly?

**Comment by sechkova** [comment]
Forgot to mention that root as a special case should be already covered in `test_updater_top_level_update.py` but probably this claim should be confirmed.

**Comment by jku** [comment]
oh yeah, validating that local metadata is actually used when downloaded metadata is verified (the second case in the issue description) is tricky. I think doing just the first case in the issue description (validate that correct files are read from disk the correct number of times) is a good first step -- so that would be roughly steps 1-3 in your comment. I would argue it's mostly just augmenting the test function you wrote, but a new one is fine too

**Comment by MVrachev** [comment]
Which steps of the @sechkova comment should be tested to consider this issue closed?
If it is as @jku said - steps 1 to 3 do we want to create a separate issue for the rest?

**Comment by MVrachev** [comment]
> What I imagined while writing the issue is potentially having a class that tests all weird cases of loading cached metadata. This would mean at least:
> 
> Step 1: make a successful update of valid metadata which stores it in cache (by calling preferably Updater.get_targetinfo or in some special case Updater.refresh)

@sechkova why it's preferable to call `Updater.get_targetinfo ` instead of directly calling `refresh`?
Maybe I am missing some context here.

**Comment by sechkova** [comment]
> @sechkova why it's preferable to call `Updater.get_targetinfo ` instead of directly calling `refresh`?
> Maybe I am missing some context here.

Seems like I've had in mind refreshing both top-levels and delegated roles metadata when writing this. Updater.get_targetinfo calls refresh internally.

**Comment by MVrachev** [comment]
> > @sechkova why it's preferable to call `Updater.get_targetinfo ` instead of directly calling `refresh`?
> > Maybe I am missing some context here.
> 
> Seems like I've had in mind refreshing both top-levels and delegated roles metadata when writing this. Updater.get_targetinfo calls refresh internally.

As for steps 1, 2 and 4 implemented in https://github.com/theupdateframework/python-tuf/pull/1707 we need to test that only expired top-level roles have been used to verify metadata files, then I think we can just use refresh here.

**Comment by sechkova** [comment]
> * test that an expired targets when loaded from cache  is not stored as final but is used to verify the new downloaded targets

Yes, I think I'm wrong in this sentence, I edited the comment. Expired targets is always discarded, expired delegations too.

**Comment by MVrachev** [comment]
> > * test that an expired targets when loaded from cache  is not stored as final but is used to verify the new downloaded targets
> 
> Yes, I think I'm wrong in this sentence, I edited the comment. Expired targets is always discarded, expired delegations too.

Looking at the code and testing it locally, if a local targets has expired we throw a `ExpiredMetadataError` https://github.com/theupdateframework/python-tuf/blob/bdf1cbbbb56839cd99b965fedae2661ed873c4a2/tuf/ngclient/_internal/trusted_metadata_set.py#L459
which inherits a `RepositoryError` and we will catch this exception in line 394 from here: https://github.com/theupdateframework/python-tuf/blob/bdf1cbbbb56839cd99b965fedae2661ed873c4a2/tuf/ngclient/updater.py#L394 Then, the update will continue and try to download and update the next version.

In conclusion: if a local targets has expired the update process won't stop, but the local targets won't be verified and the update will try to fetch and verify the next version without using any information from the old expired targets file.
